// modules/display/display.js
'use strict';
/**
 * Song display
 */

let props = Object.defineProperties,
  Progress = require('modules/display/progress'),
  EventEmitter = require('events'),
  inherits = require('util').inherits;

function Display() {
  props(this, {
    _mainTitle: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    },
    _subTitle: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    },
    _artwork: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    },
    _progress: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: new Progress()
    }
  })

  EventEmitter.call(this);
}

inherits(Display, EventEmitter);

Display.prototype.reset = function() {
  this.setMainTitle(null);
  this.setSubTitle(null);
};

Display.prototype.setMainTitle = function(value) {
  this._mainTitle = value;
  this.emit('update-main-title', value);
  return this;
};

Display.prototype.getMainTitle = function() {
  return this._mainTitle;
};

Display.prototype.setSubTitle = function(value) {
  this._mainTitle = value;
  this.emit('update-main-title', value);
  return this;
};

Display.prototype.getSubTitle = function() {
  return this._subTitle;
};

Display.prototype.setArtwork = function(value) {
  this._artwork = value;
  this.emit('update-artwork', value);
  return this;
};

Display.prototype.getArtwork = function() {
  return this._artwork;
};

Display.prototype.getProgressBar = function() {
  return this._progress;
};

Display.prototype.setValue = function(key, value) {
  let el = $('[data-bind="' + key + '"]');
  if (el && el.length > 0) {
    switch (el.prop('nodeName')) {
      case 'DIV':
      case 'SPAN':
      case 'TEXTAREA':
      default:
        el.html(value);
        break;
      case 'IMG':
        el.attr('src', value);
        break;
      case 'A':
        el.attr('href', value);
        break;
      case 'INPUT':
        el.val(value);
        break;
      case 'SELECT':
        el.val(value);
        let option = el.find('option[value="' + value + '"]');
        if (option && option.length > 0) {
          el.find('option:selected').removeAttr('selected');
          option.attr('selected', true);
        }
        break;
    }
    return true;
  }
  return false;
};

Display.prototype.getValue = function(key) {
  let el = $('[data-bind="' + key + '"]'),
    value = null;
  if (el && el.length > 0) {
    switch (el.prop('nodeName')) {
      case 'DIV':
      case 'SPAN':
      case 'TEXTAREA':
      default:
        value = el.html();
        break;
      case 'INPUT':
        value = el.val();
        break;
      case 'SELECT':
        let option = el.find('option:selected');
        if (option && option.length > 0) {
          value = option.attr('value');
        } else {
          value = el.val();
        }
        break;
    }
  }
  return value;
};

Display.prototype.addClass = function(key, value) {
  let el = $('[data-bind-class="' + key + '"]');
  if (el && el.length > 0) {
    el.addClass(value);
  }
};

Display.prototype.removeClass = function(key, value) {
  let el = $('[data-bind-class="' + key + '"]');
  if (el && el.length > 0) {
    el.removeClass(value);
  }
};

Display.prototype.timeString = function(time) {
  time = parseInt(time, 10);

  let output = '00:00',
    hour = 60 * 60,
    hours = 0,
    minutes = 0,
    seconds = 0;

  if (time === 0) {
    return output;
  }

  function _leadingZero(value) {
    return value < 10 ? '0' + value : value;
  }

  hours = _leadingZero(Math.floor(time / hour));
  minutes = _leadingZero(Math.floor(time % hour / 60));
  seconds = _leadingZero(Math.floor(time % hour % 60));

  output = '';
  if (hours > 0) {
    output += `${hours}:`;
  }
  output += `${minutes}:${seconds}`;

  return output;
};

module.exports = Display;
