// modules/display/playlist-view.js
'use strict';
/**
 * Song display
 */

let props = Object.defineProperties,
  Collection = require('classes/collection');

function tagOpen(name, attrs) {
  let _attrs = '';
  Object.keys(attrs = attrs || {}).forEach(function(key) {
    _attrs += ` ${key}="${attrs[key]}"`;
  });
  return `<${name}${_attrs}>`;
}

function tagClose(name) {
  return `</${name}>`;
}

function PlaylistView() {
  props(this, {
    _columns: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: Object.create(null)
    },
    data: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: new Collection()
    }
  });
}

PlaylistView.prototype.getColumns = function() {
  return this._columns;
};

PlaylistView.prototype.setColumns = function(columns) {
  columns = columns || {};
  if (typeof columns !== 'object') {
    throw new TypeError('Invalid arguments');
  }
  let keys = Object.keys(columns),
    _this = this;
  if (keys.length === 0) {
    return false;
  }
  keys.forEach(function(key) {
    _this.setColumn(key, columns[key]);
  });
  return true;
};

PlaylistView.prototype.getColumn = function(key) {
  return this.hasColumn(key) ? this._columns[key] : null;
};

PlaylistView.prototype.setColumn = function(key, callback) {
  if (typeof callback === 'function') {
    this._columns[key] = callback;
    return true;
  }
  return false;
};

PlaylistView.prototype.removeColumn = function(key) {
  if (this.hasColumn(key)) {
    this._columns[key] = null;
    delete this._columns[key];
    return true;
  }
  return false;
};

PlaylistView.prototype.hasColumn = function(key) {
  return typeof this._columns[key] === 'function';
};

PlaylistView.prototype.render = function() {
  let _this = this,
    columns = this.getColumns(),
    ckeys = Object.keys(columns),
    data = this.data.getRaw(),
    dkeys = Object.keys(data),
    headers = [],
    content = [],
    rowColumns = null,
    attrs = {};

  ckeys.forEach(function(title) {
    headers.push(`${tagOpen('div', { class: 'table-column table-head-column' })}${title}${tagClose('div')}`);
  });

  dkeys.forEach(function(index) {
    rowColumns = [];
    ckeys.forEach(function(ckey) {
      rowColumns.push(`${tagOpen('div', { class: 'table-column' })}
        ${columns[ckey].call(_this, data[index])}
        ${tagClose('div')}`);
    });

    attrs = {'data-track-id': data[index].get('id'), class: 'table-row'};
    content.push(`${tagOpen('div', attrs)}${rowColumns.join('')}${tagClose('div')}`);
  });

  return [
    tagOpen('div', { class: 'table' }),
      tagOpen('div', { class: 'table-head' }),
        tagOpen('div', { class: 'table-row' }),
          headers.join(''),
        tagClose('div'),
      tagClose('div'),
      tagOpen('div', { class: 'table-body' }),
        content.join(''),
      tagClose('div'),
    tagClose('div')
  ].join('');
};

module.exports = PlaylistView;
