// classes/app.js
'use strict';

let util = require('util'),
  remote = require('remote'),
  App = remote.require('app'),
  Menu = remote.require('menu'),
  MenuItem = remote.require('menu-item'),
  extend = require('utils/extend'),
  EventEmitter = require('events'),
  Options = require('classes/options'),
  ModulesCollection = require('collections/modules'),
  props = Object.defineProperties,
  defaultOptions;

const APP_NAME = App.getName();

defaultOptions = {
  client_id: null,
  redirect_uri: 'connect://soundcloud'
};

function SoundWarp() {
  props(this, {
    'modules': {
      configurable: false,
      enumerable: true,
      writable: true,
      value: new ModulesCollection()
    }
  });

  EventEmitter.call(this);
  Options.call(this);
}

util.inherits(SoundWarp, EventEmitter);
extend(SoundWarp.prototype, Options.prototype);

SoundWarp.prototype.log = function() {
  return this.getOption('flags').DEBUG && window.console && console.log.apply(console, Array.prototype.slice.call(arguments));
};

SoundWarp.prototype.getMenuTemplate = function() {
  let _this = this;
  return [
    { // 0
      label: APP_NAME,
      submenu: [
        {
          label: `About ${APP_NAME}`,
          role: 'about'
        },
        {
          type: 'separator'
        },
        {
          label: 'Services',
          role: 'services',
          submenu: []
        },
        {
          type: 'separator'
        },
        {
          label: `Hide ${APP_NAME}`,
          accelerator: 'Command+H',
          role: 'hide'
        },
        {
          label: 'Hide Others',
          accelerator: 'Command+Shift+H',
          role: 'hideothers'
        },
        {
          label: 'Show All',
          role: 'unhide'
        },
        {
          type: 'separator'
        },
        {
          label: 'Quit',
          accelerator: 'Command+Q',
          click: function() {
            App.quit();
          }
        },
      ]
    },
    {// 1
      label: 'Account',
      submenu: [
        {
          label: 'Connect',
          enabled: !_this.client.isConnected(),
          click: function() {
            $('[data-action="sw.client.connect"]').trigger('click');
          }
        },
        {
          label: 'Disconnect',
          enabled: _this.client.isConnected(),
          click: function() {
            $('[data-action="sw.client.disconnect"]').trigger('click');
          }
        }
      ]
    },
    {// 2
      label: 'Window',
      role: 'window',
      submenu: [
        {
          label: 'Minimize',
          accelerator: 'CmdOrCtrl+M',
          role: 'minimize'
        },
        {
          type: 'separator'
        },
        {
          label: 'Bring All to Front',
          role: 'front'
        }
      ]
    }
  ];
};

SoundWarp.prototype.attachMenu = function(template) {
  let _this = this;
  template = template || this.getMenuTemplate();
  Menu.setApplicationMenu(Menu.buildFromTemplate(template));
};

SoundWarp.prototype.registerModule = function(name, module) {
  let instance;

  if (typeof module === 'object') {
    instance = module;
  } else {
    instance = module.call(this);
  }

  this.modules.push(name, instance);

  Object.defineProperty(this, name, {
    configurable: false,
    enumerable: true,
    set: function(value) {
      // void
    },
    get: function() {
      return this.modules.get(name);
    }
  });

  this.emit('register-module', {
    name: name,
    module: this[name],
    constructor: module
  });
  return this;
};

SoundWarp.prototype.registerModules = function(modules) {
  let _this = this;
  Object.keys(modules).forEach(function(name) {
    _this.registerModule(name, modules[name]);
  });
  return this;
};

module.exports = SoundWarp;
