// modules/client/index.js
'use strict';
/**
 * Client
 */

let props = Object.defineProperties,
  EventEmitter = require('events'),
  inherits = require('util').inherits,
  qs = require('querystring'),
  has = Object.prototype.hasOwnProperty;

require(__dirname + '/soundcloud');

function Client() {
  props(this, {
    _options: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    },
    URL_TOKEN: {
      configurable: false,
      enumerable: false,
      writable: false,
      value: 1
    },
    URL_CLIENT: {
      configurable: false,
      enumerable: false,
      writable: false,
      value: 2
    }
  });

  EventEmitter.call(this);
}

inherits(Client, EventEmitter);

Client.prototype.setOptions = function(options) {
  options = options || {};

  let keys = Object.keys(options),
    _this = this;

  if (keys.length > 0) {
    Object.keys(defaultOptions).forEach(function(key) {
      _this.setOption(key, keys[key] ? options[key] : defaultOptions[key]);
    });
  }
  return this;
};

Client.prototype.setOption = function(key, value) {
  this._options[key] = value;
  this.emit('update-option', {
    name: key,
    value: value
  });
  return this;
};

Client.prototype.connect = function(callback) {
  let _this = this,
    emptyFn = function() {};

  window.onmessage = function(event) {
    if (event.origin === 'connect://') {
      try {
        let data = JSON.parse(event.data);
        if (has.call(data, 'location')) {
          let index = data.location.indexOf('?') + 1,
            query = qs.parse(data.location.substring(index)),
            state = query.state.split('#'),
            accessToken = state[1].substring(state[1].indexOf('=') + 1);
          _this.initialize({
            accessToken: accessToken
          });
          SC._connectWindow.close();
          window.onmessage = emptyFn;
          return callback();
        }
      } catch (err) {
        window.onmessage = emptyFn;
        return callback(err);
      }
    }
  };

  SC.connect(emptyFn);
};

Client.prototype.initialize = function(optionsOrCallback) {
  optionsOrCallback = optionsOrCallback || {};
  let config = {};
  if (typeof optionsOrCallback === 'object') {
    config.access_token = optionsOrCallback;
  } else {
    config = optionsOrCallback;
  }
  return SC.initialize(config);
};

Client.prototype.get = function(path, params, callback) {
  return SC.get(path, params, callback);
};

Client.prototype.post = function(path, params, callback) {
  return SC.post(path, params, callback);
};

Client.prototype.delete = function(path, params, callback) {
  return SC.delete(path, params, callback);
};

Client.prototype.url = function(url, type) {
  type = type || this.URL_CLIENT;
  url = url + (url.indexOf('?') > -1 ? '&' : '?');
  switch (type) {
    case 1:
    default:
      url = url + 'client_id=' + SC.options.client_id;
      break;
    case 2:
      url = url + 'oauth_token=' + SC.accessToken;
      break;
    case 3:
      url = url + 'client_id=' + SC.options.client_id + '&' +
        'oauth_token=' + SC.accessToken;
      break;
  }
  return url;
};

module.exports = Client;
