// utils/dom-utils.js
'use strict';
/**
 * DOM utils
 */

function __create_getter(attr) {
  return function(selector) {
    let el = $(selector);
    if (el.length > 0) {
      return el.attr(attr);
    }
    return null;
  };
};

function __create_setter(attr) {
  return function(selector, value) {
    let el = $(selector);
    if (el.length > 0) {
      el.attr(attr, value);
      el.removeClass('hidden');
      return el;
    }
    return null;
  };
};

module.exports = {
  getlet: __create_getter('data-let'),
  setlet: __create_setter('data-let'),
  getSrc: __create_getter('data-src'),
  setSrc: __create_setter('data-src'),
  getAttr: function(selector, attr) {
    let el = $('[data-' + attr + '="' + selector + '"]');
    if (el.length > 0) {
      return el.attr(attr);
    }
    return null;
  },
  setAttr: function(selector, attr, value) {
    let el = $('[data-' + attr + '="' + selector + '"]');
    if (el.length > 0) {
      switch (attr) {
        case 'width':
        case 'height':
          el[attr](value);
          break;
        case 'value':
          el.val(value);
          break;
        default:
          el.attr(attr, value);
          break;
      }
      el.removeClass('hidden');
      return el;
    }
    return null;
  },
};
