// modules/display.js
'use strict';
/**
 * Song display
 */

let props = Object.defineProperties,
  EventEmitter = require('events'),
  inherits = require('util').inherits;

function Display() {
  props(this, {
    _mainTitle: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    },
    _subTitle: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    },
    _artwork: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    }
  })

  EventEmitter.call(this);
}

inherits(Display, EventEmitter);

Display.prototype.reset = function() {
  this.setMainTitle(null);
  this.setSubTitle(null);
};

Display.prototype.setMainTitle = function(value) {
  this._mainTitle = value;
  this.emit('update-main-title', value);
  return this;
};

Display.prototype.getMainTitle = function() {
  return this._mainTitle;
};

Display.prototype.setSubTitle = function(value) {
  this._mainTitle = value;
  this.emit('update-main-title', value);
  return this;
};

Display.prototype.getSubTitle = function() {
  return this._subTitle;
};

Display.prototype.setArtwork = function(value) {
  this._artwork = value;
  this.emit('update-artwork', value);
  return this;
};

Display.prototype.getArtwork = function() {
  return this._artwork;
};

module.exports = Display;
