// modules/profile.js
'use strict';
/**
 * Song display
 */

let props = Object.defineProperties,
  ProfileCollection = require('collections/profile'),
  Artwork = require('modules/player/artwork');

function Profile() {
  props(this, {
    data: {
      configurable: false,
      enumerable: true,
      writable: true,
      value: new ProfileCollection()
    },
    _avatar: {
      configurable: false,
      enumerable: false,
      writable: true,
      value: null
    }
  })
}

Profile.prototype.reset = function() {
  this.data.clear();
  return this;
};

Profile.prototype.getData = function() {
  return this.data;
};

Profile.prototype.setData = function(data) {
  let _this = this,
    keys = Object.keys(data);

  if (keys.length === 0) {
    return false;
  }

  keys.forEach(function(key) {
    _this.set(key, data[key]);
  });

  return true;
};

Profile.prototype.get = function(key) {
  return this.data.get(key, null);
};

Profile.prototype.set = function(key, value) {
  this.data.set(key, value);
  return this;
};

Profile.prototype.isEmpty = function() {
  return this.data.size() === 0;
};

Profile.prototype.getAvatarUrl = function(size) {
  if ((this._avatar instanceof Artwork) === false) {
    this._avatar = new Artwork(this.data.get('avatar_url'));
  }
  return this._avatar.getUrl(size);
};

module.exports = Profile;
